File: NOI/2004/day1/lib/dune_lib.cpp
#include <fstream>
#include <string>

using namespace std;

ifstream fin("dune.in");
ofstream fout;

static const int MaxN = 101;
static const int MaxM = 4001;

static int n, m, nowpos, nowroad, Count, score;
static bool stone;
static bool havestone[MaxN];
static bool cave, channel;
static int g[MaxN][MaxM], h[MaxN][MaxM];
static int d[MaxN];
static bool Initialized = false, isInTest = false;

static void Error(string message)
{
	if (!isInTest) fout << message << endl;
	else fout << -1 << endl;
    exit(0);
}

void init()
{
	if (Initialized) {
		Error("You have called init() twice!");
	}
    Initialized=true;
    fin >> n;
    if (n == 0) {
    	isInTest = true;
    	fin >> n >> m;
    	fout.open("dune.out");
    }
    else fout.open("dune.log");
	if (!isInTest) fout << "init()" << endl;
    m = 0;
    int i, j, k;
    for (i = 1; i <= n; i++) {
        fin >> d[i];
        m += d[i];
        for (j = 0; j < d[i]; j++) {
            fin >> g[i][j];
        }
    }
    m /= 2;
    for (i = 1; i <= n; i++) {
        for (j = 0; j < d[i]; j++) {
            for (k = 0; k < d[g[i][j]]; k++) {
                if (g[g[i][j]][k] == i) break;
            }
            h[i][j] = k;
        }
    }
    nowpos = 1;
    nowroad = 0;
    stone = true;
    memset(havestone, 0, sizeof(havestone));
    Count = 0;
    score = 0;
    cave = channel = false;
}

void put_sign()
{
	if (!isInTest) fout << "put_sign()" << endl;
    if (!Initialized) Error("You must call init() first!");
    if (!stone) Error("Error in put_sign()!");
    stone = false;
    havestone[nowpos] = true;
}

void take_sign()
{
	if (!isInTest) fout << "take_sign()" << endl;
    if (!Initialized) Error("You must call init() first!");
    if (!havestone[nowpos]) Error("Error in take_sign()!");
    stone = true;
    havestone[nowpos] = false;
}

void walk(int k)
{
	if (!isInTest) fout << "walk(" << k << ")" << endl;
    if (!Initialized) Error("You must call init() first!");
    int nextpos = g[nowpos][(nowroad + k) % d[nowpos]];
    nowroad = h[nowpos][(nowroad + k) % d[nowpos]];
    nowpos = nextpos;
    Count++;
}

void look(int &roadnum, bool &stone)
{
	if (!isInTest) fout << "look(d, sign), returned with d=" << d[nowpos] << ", sign=" << havestone[nowpos] << endl;
    if (!Initialized) Error("You must call init() first!");
    roadnum = d[nowpos];
    stone = havestone[nowpos];
}

void report(int ncave, int nchannel)
{
	if (!isInTest) fout << "report(" << ncave << ", " << nchannel << ")" << endl;
    if (!Initialized) Error("You must call init() first!");
    if (!isInTest) {
	    if (ncave == n) fout << "The number of caves is correct!" << endl;
	    else fout << "The number of caves is NOT correct!" << endl;
	    if (nchannel == m) fout << "The number of channels is correct!" << endl;
	    else fout << "The number of channels is NOT correct!" << endl;
	    fout << "You have walked " << Count << " time(s)!" << endl;
	}
	else {
		fout << Count << endl;
		fout << ncave << ' ' << nchannel << endl;
	}
    exit(0);
}

File: NOI/2004/day1/lib/dune_lib.pas
unit dune_lib;

interface

procedure init;
procedure put_sign;
procedure take_sign;
procedure walk(i:longint);
procedure look(var roadnum:longint; var stone:boolean);
procedure report(ncave,nchannel:longint);

implementation

const
	MaxN=100;
	MaxM=4000;
	
var
	n,m,nowpos,nowroad,Count,score:longint;
	initialized, isInTest:boolean;
	stone,cave,channel:boolean;
	havestone:array [1..MaxN] of boolean;
	g,h:array [1..MaxN,0..MaxM-1] of longint;
	d:array [1..MaxN] of longint;

procedure error(s:string);
begin
	if (not isInTest) then writeln(s)
	else writeln(-1);
	close(input);
	close(output);
	halt;
end;
	
procedure init;
var
	i,j,k:longint;
begin
	if (initialized) then error('You have called init twice!');
	initialized:=true;
	assign(input,'dune.in');
	reset(input);
	read(n);
	isInTest:=false;
	if (n=0) then
	begin
		read(n, m);
		isInTest:=true;
	end;
	if (isInTest) then assign(output, 'dune.out')
	else assign(output, 'dune.log');
	rewrite(output);
    if (not isInTest) then writeln('init');
	m:=0;
	for i:=1 to n do
	begin
		read(d[i]);
		inc(m,d[i]);
		for j:=0 to d[i]-1 do read(g[i,j]);
	end;
	m:=m div 2;
	for i:=1 to n do
		for j:=0 to d[i]-1 do
		begin
			for k:=1 to d[g[i,j]] do
				if g[g[i,j],k]=i then break;
			h[i,j]:=k;
		end;
	nowpos:=1;
	nowroad:=0;
	stone:=true;
	fillchar(havestone,sizeof(havestone),0);
        Count:=0;
	score:=0;
	cave:=false;
	channel:=false;
end;

procedure put_sign;
begin
	if (not isInTest) then writeln('put_sign');
	if (not initialized) then error('You must call init first');
	if (not stone) then error('Error in put_sign');
	stone:=false;
	havestone[nowpos]:=true;
end;

procedure take_sign;
begin
	if (not isInTest) then writeln('take_sign');
	if (not initialized) then error('You must call init first');
	if (not havestone[nowpos]) then error('Error in take_sign');
	stone:=true;
	havestone[nowpos]:=false;
end;

procedure walk(i:longint);
var
	nextpos:longint;
begin
	if (not isInTest) then writeln('walk(', i, ')');
	if (not initialized) then error('You must call init first');
	nextpos:=g[nowpos,(nowroad+i) mod d[nowpos]];
	nowroad:=h[nowpos,(nowroad+i) mod d[nowpos]];
	nowpos:=nextpos;
	inc(Count);
end;

procedure look(var roadnum:longint; var stone:boolean);
begin
	if (not isInTest) then writeln('look(d, sign) returned with d=', d[nowpos], ', sign=', havestone[nowpos]);
	if (not initialized) then error('You must call init first');
	roadnum:=d[nowpos];
	stone:=havestone[nowpos];
end;

procedure report(ncave,nchannel:longint);
begin
	if (not isInTest) then writeln('report(', ncave, ', ', nchannel, ')');
	if (not initialized) then error('You must call init first');
	if (not isInTest) then
	begin
		if ncave=n then writeln('The number of caves is correct!')
		else writeln('The number of caves is NOT correct!');
		if nchannel=m then writeln('The number of channels is correct!')
		else writeln('The number of channels is NOT correct!');
        writeln('You have walked ',Count,' time(s)!');
    end
    else begin
    	writeln(count);
    	writeln(ncave, ' ', nchannel);
    end;
    close(input);
    close(output);
    halt;
end;

begin
  initialized := false;
end.

